src/Algebra/Graph/LabelledAdjacencyMap.hs

```diff
@@ -0,0 +1,780 @@
+-- Complexity: /O((n + m) * log(n))/ time.
+--
+-- @
+-- gmap f 'empty'      == 'empty'
+-- gmap f ('vertex' x) == 'vertex' (f x)
+-- gmap f ('edge' x y) == 'edge' (f x) (f y)
+-- gmap id           == id
+-- gmap f . gmap g   == gmap (f . g)
+-- @
+gmap
+        :: (Ord a, Ord b)
+        => (a -> b)
+        -> LabelledAdjacencyMap a e
+        -> LabelledAdjacencyMap b e
+gmap f =
+        LAM
+                . Map.map (Map.mapKeys f)
+                . Map.mapKeysWith Map.union f
+                . labelledAdjacencyMap
+                
+-- | Transform a graph by applying a function to each of its edge labels. This is
+-- similar to @Functor@'s 'fmap' but can be used with non-fully-parametric
+-- 'LabelledAdjacencyMap'.
+-- Complexity: /O((n + m) * log(n))/ time.
+--
+-- @
+-- gemap f 'empty'      == 'empty'
+-- gemap f ('vertex' x) == 'vertex' (f x)
+-- gemap f ('edge' x y) == 'edge' (f x) (f y)
+-- gemap id           == id
+-- gemap f . gmap g   == gmap (f . g)
+-- @
+gemap :: (e -> e') -> LabelledAdjacencyMap a e -> LabelledAdjacencyMap a e'
+gemap f = LAM . Map.map (Map.map f) . labelledAdjacencyMap
+
+-- | Construct the /induced subgraph/ of a given graph by removing the
+-- vertices that do not satisfy a given predicate.
+-- Complexity: /O(m)/ time, assuming that the predicate takes /O(1)/ to
+-- be evaluated.
+--
+-- @
+-- induce (const True ) x      == x
+-- induce (const False) x      == 'empty'
+-- induce (/= x)               == 'removeVertex' x
+-- induce p . induce q         == induce (\\x -> p x && q x)
+-- 'isSubgraphOf' (induce p x) x == True
+-- @
+induce
+        :: (a -> Bool)
+        -> LabelledAdjacencyMap a e
+        -> LabelledAdjacencyMap a e
+induce p =
+        LAM
+                . Map.map (Map.filterWithKey (\k _ -> p k))
+                . Map.filterWithKey (\k _ -> p k)
+                . labelledAdjacencyMap
+
+
+
+-- | Compute the /depth-first search/ forest of a graph.
+--
+-- @
+-- 'forest' (dfsForest $ 'edge' 1 1)         == 'vertex' 1
+-- 'forest' (dfsForest $ 'edge' 1 2)         == 'edge' 1 2
+-- 'forest' (dfsForest $ 'edge' 2 1)         == 'vertices' [1, 2]
+-- 'isSubgraphOf' ('forest' $ dfsForest x) x == True
+-- dfsForest . 'forest' . dfsForest        == dfsForest
+-- dfsForest ('vertices' vs)               == map (\\v -> Node v []) ('Data.List.nub' $ 'Data.List.sort' vs)
+-- 'dfsForestFrom' ('vertexList' x) x        == dfsForest x
+-- dfsForest $ 3 * (1 + 4) * (1 + 5)     == [ Node { rootLabel = 1
+--                                                 , subForest = [ Node { rootLabel = 5
+--                                                                      , subForest = [] }]}
+--                                          , Node { rootLabel = 3
+--                                                 , subForest = [ Node { rootLabel = 4
+--                                                                      , subForest = [] }]}]
+-- @
+dfsForest :: (Ord a) => LabelledAdjacencyMap a e -> Forest a
```

Comment:
If you make `LabelledAdjacencyMap` an instance of `ToGraph` then there is no need to reimplement all this -- just use the default `ToGraph` implementations.